### PR TITLE
Hook up the "Add Assignment" dialog to create an assignment 

### PR DESCRIPTION
**Pull request by payten**
Hi @steveswinsburg, I've had a crack at hooking up the add assignment dialog to create an assignment via the service and also display exceptions as error messages in the dialog.

I noticed that the categories are not sticking? So perhaps I've missed something there.

Also, I can't seem to get around the unload browser prompt when leaving the page upon successfully adding an assignment.  Is there a way to suppress it?

Anyway, hopefully this is useful.  Let me know if you'd like me to look into anything in particular.  Otherwise I'll keep prodding at holes I find ;)

### CODE REVIEW

File: tool/src/java/org/sakaiproject/gradebookng/tool/pages/BasePage.java [hunk 1]
```diff
@@ -38,7 +38,7 @@
 	private static final Logger log = Logger.getLogger(BasePage.class); 
 	
 	@SpringBean(name="org.sakaiproject.gradebookng.business.GradebookNgBusinessService")
-	protected GradebookNgBusinessService businessService;
+	public GradebookNgBusinessService businessService;
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/pages/BasePage.java]: This shouldn't be public. It needs to be injected into the other panels.

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java [hunk 2]
```diff
@@ -14,14 +27,82 @@
 
 	private static final long serialVersionUID = 1L;
 
-	public AddGradeItemPanel(String id) {
+	public AddGradeItemPanel(String id, List<CategoryDefinition> categories, final ModalWindow window) {
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java]: Have a look at the other panels. If you inject the business service in you can do this logic inside there. then have a getter in the gradebook page for the modal window.  Passing stuff into the constructor isnt safe for this.

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java [hunk 3]
```diff
@@ -14,14 +27,82 @@
...
 
+    AjaxButton cancel = new AjaxButton("cancel") {
+      private static final long serialVersionUID = 1L;
+      @Override
+      public void onSubmit(AjaxRequestTarget target, Form<?> form) {
+        window.close(target);
+      }
+    };
+    form.add(cancel);
+    add(form);
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java]: Some weird spacing happening here?

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java [hunk 4]
```diff
@@ -20,29 +23,18 @@
 
 	private static final long serialVersionUID = 1L;
 
-    public AddGradeItemPanelContent(String id, GbAssignmentModel gbAssignmentModel) {
+    public AddGradeItemPanelContent(String id, GbAssignmentModel gbAssignmentModel, List<CategoryDefinition> categories) {
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java]: As above, inject the service and look this up

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java [hunk 5]
```diff
@@ -7,9 +7,12 @@
 import org.apache.wicket.model.Model;
 import org.apache.wicket.model.PropertyModel;
 import org.sakaiproject.gradebookng.tool.model.GbAssignmentModel;
+import org.sakaiproject.service.gradebook.shared.CategoryDefinition;
 
 import java.util.Arrays;
 import java.util.List;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java]: erroneous import?

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/importExport/CreateGradeItemStep.java [hunk 6]
```diff
@@ -106,7 +108,8 @@ public void onSubmit() {
 
         form.add(new Label("createItemHeader", new StringResourceModel("importExport.createItem.heading", this, null, step, importWizardModel.getTotalSteps())));
 
-        AddGradeItemPanelContent gradePanelContent = new AddGradeItemPanelContent("subComponents", gbAssignmentModel);
+        List<CategoryDefinition> categories = ((BasePage) getPage()).businessService.getGradebookCategories();
+        AddGradeItemPanelContent gradePanelContent = new AddGradeItemPanelContent("subComponents", gbAssignmentModel, categories);
```
Comment (steveswinsburg) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/importExport/CreateGradeItemStep.java:110]: As above re refactor, these two lines not required.

**Comment by steveswinsburg** [comment]
Hey Payten, you need to inject the service into the panel to save passing data around into constructors which isn't safe. Have a look at the other panels/windows. Also, there is a pattern for declaring the modal in the gradebookpage and then providing a getter, can you make this do the same?

To remove the unload confirmations, set:
`modalWindow.showUnloadConfirmation(false);`

For the categories, I had a look, the category is not included in the GradebookService call to add assignment, nor to update :-1: 
I'll have to send a PR for that :rage:

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java (hunk 4 above)
Comment (payten) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java]: Gotcha, thanks!

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java (hunk 5 above)
Comment (payten) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanelContent.java]: Thanks!  Yep, was going to pull in some Java 8 stream stuff but the compiler wouldn't let me get away with it.  Will drop them out.

File: tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java (hunk 3 above)
Comment (payten) [inline tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddGradeItemPanel.java]: Ah crap yeah.  Any preference what tabs/formatting we go with? I've found I have to change the IDE settings file to file from soft to hard tabs, 2-4 spaces.

**Comment by payten** [comment]
Thanks for the help @steveswinsburg.  Those changes are all done.

**Comment by steveswinsburg** [comment]
See https://jira.sakaiproject.org/browse/SAK-29734 for the API change that should make categories work.

**Pull request merged**